### ISSUE TITLE
twisted.internet.task.react should actually exit python, using appropriate exit codes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6159]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6159|
|Type|enhancement|
|Created|2012-10-24 11:52:19Z|
|Branch|https://github.com/twisted/twisted/tree/react-exit-6159|

Exit codes tend to be more useful for command-line programs (the use case for `react`) than servers. As such, it seems like it'd be good if `react` actually exited with an exit code after the reactor is done.

  * By default exit code should be 0.
  * If an exception is raised/result Deferred is fired with errback, it should be 1.
  * If a `SystemExit` exception is raised or passed as errback of result Deferred, exit with the code specified by the `SystemExit`.

This is an API change from current version, so we should decide if we're doing this before the next release since the API is still unreleased and therefore does not require backwards compat.

<details><summary>Searchable metadata</summary>

```
trac-id__6159 6159
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Twisted_12_3 Twisted-12.3
branch__branches_react_exit_6159 branches/react-exit-6159
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1351079539000000 1351079539000000
changetime__1355417120000000 1355417120000000
version__None None
owner__therve therve
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [36479]) Branching to 'react-exit-6159'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

OK, this is ready for review. This was definitely harder than I expected, and the refactoring is certainly big. One questionable thing I changed is due to a problem I encountered when testing manually, illustrated by the new test_synchronousStop. I'm not sure why we support the user calling reactor.stop himself, though, as he doesn't call reactor.run.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

Thanks!

I think supporting the user calling `reactor.stop` themselves is important in order to make `react` generally useful, rather than something that will randomly break depending on which particular code you happen to be invoking (directly or indirectly - in the glorious future lacking a global reactor, this may be less of an issue, but for now it's too easy for any random code to stop the reactor on its own).

  1. The news fragment should be a misc, since 3270.feature already describes `react`.
  1. I'm surprised to see the code changed to depend on `reactor.running`.  I think `running` is an implementation detail that we very grudgingly exposed, with fairly complicated and subtle semantics (for example, I don't think `_FakeReactor` is faithfully reproducing the transition to `False`, which should happen before the first application-added _during shutdown_ hook runs, but in `_FakeReactor` it happens after the last _during shutdown_ hook runs - which I think means the `reactor.stop()` support is actually broken against a real reactor) to satisfy people who were already using it or who find themselves in some particularly terrible situation - not something to actually use in our own code (at least not our code outside of the reactor implementation).  I guess the documentation could be more clear on that point...  Still, I wonder what was wrong with the previous approach, and whether that could be fixed instead of introducing this `running` usage.
  1. It might be nice if all the unit tests made an assertion about the `SystemExit`'s code (most do, but `test_runsUntilSyncCallback`, `test_runsUntilSyncErrback`, `test_arguments`, `test_defaultReactor` don't).
  1. Can you file a ticket for adding unit tests for `_FakeReactor`?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Thanks, everything fixed and [#6200](https://github.com/twisted/twisted/issues/6200) opened. The change to running was probably due to some confusion about _FakeReactor behavior, and it was reverted.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve<br>@exarkun set status to `new`|
|-|-|

Thanks.

  1. One case is not well handled with the latest version of the code - the case where something calls `reactor.stop()` and the `Deferred` returned by the `main` function does _not_ quickly (ie, before `reactor.run` returns) fire.  This is easy demonstrated with a main function that returns a `Deferred` that _never_ fires:

```python
from twisted.internet.task import react
from twisted.internet import defer

def main(reactor):
    return defer.Deferred()

react(main, [])
```

Run this and hit Ctrl-C.  The result is an `IndexError` while trying to determine the proper exit code to use.

Once that case is fixed and buildbot is green, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [36544]) Merge react-exit-6159

Author: therve
Reviewer: exarkun
Fixes: [#6159](https://github.com/twisted/twisted/issues/6159)

Change twisted.internet.task.react to exit the python process with non-0 exit
code when the deferred returned failed.
